Pydevsup

gustavosr8 opened the pull request.

gustavosr8 committed:
Big bang - pyDevSup

gustavosr8 committed:
Correct link process to writing operations

gustavosr8 committed:
Add support to epics-in-docker.

gustavosr8 committed:
Add README

gustavosr8 committed:
Add resource usage monitor with `htop`

gustavosr8 committed:
Set the Kicks PVs to have a passive scan and be updated by Current PVs

gustavosr8 committed:
Read all correctors in a given section.

gustavosr8 committed:
Remove hardcoded dipole strength and add PV reading instead.

gustavosr8 committed:
Improve style and remove boilerplate comments. Also add EGUs.

gustavosr8 committed:
Add proxy to monitor link state with external PV.

gustavosr8 committed:
Add Major alarm when external PV disconnected.

gustavosr8 committed:
Move modules paths from RELEASE to RELEASE.local.

gustavosr8 committed:
Remove user specific definitions.

gustavosr8 committed:
Add PREC fields.

gustavosr8 committed:
Add `DRVL` and `DRVH` fields in `Kick-SP` based on Current ones.

The Scan is necessary to always update the converted values to follow
new energy readings, doesn't relying on updates from the Current-SP
itself.

gustavosr8 committed:
Split the `sseq` proxy.

Each time one of the records was updated, all the others were processed
because all the records were copied from the same 'sseq' proxy.

gustavosr8 committed:
Add loop for Kick and Current setpoints.

Before this, the Kick-SP updates the Current-SP, but the inverse doesn't
occurs. Now, the setpoints are always matching, no matter how was the
one written. This is done to update Kick-SP when Current-SP is updated
by someone else.

gustavosr8 committed:
Start the IOC for all crates at once.

gustavosr8 committed:
Add script to generate crate substitutions based on environment var.

This allows us to select for which crate do we want the IOC after build
it, just defining the SEC_LIST var.

gustavosr8 committed:
Add test macro prefix.

gustavosr8 committed:
Change the scan period on limit PVs to 5 seconds.

This is necessary due to an thread overrun in the .5 thread. Since the
limits doesn't changes in a high frequency, we can use a slower scan
period to handle this task.

gustavosr8 committed:
Change conversion functions to only receive necessary arguments.

gustavosr8 committed:
Use one EnergyRef-Mon PV per crate.

This must be done to avoid duplicates when several instances of this IOC
are used to different crates. So, this prevents two records with the
same name to be created.

gustavosr8 committed:
Add image build in ci.

gustavosr8 committed:
Add usage instructions.

gustavosr8 committed:
Add tagged image build in docker-compose.

ericonr commented on README.md:
```diff
@@ -0,0 +1,29 @@
+You can use the following command to run it in the background using the default
+start-up script from
+[epics-in-docker](https://github.com/cnpem/epics-in-docker). First, define the
+variable `SEC_LIST` with the list of sections to handle on this IOC instance.
```
```suggestion
variable `SEC_LIST` with the list of storage ring sectors to handle on this IOC instance.
```

ericonr commented on docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+    build:
+      context: ./
+      dockerfile: docker/Dockerfile
+      target: dynamic-link
+      labels:
+        org.opencontainers.image.source: https://github.com/cnpem/conv-fastcorrs-epics-ioc
```
I remember you tried updating to the latest epics-in-docker release. Was there any blocker for pushing it to this branch?

ericonr commented on docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+        RUNDIR: /opt/conv-fastcorrs-epics-ioc/iocBoot/iocsi_ps_conv_fastcorrs
+        BUILD_PACKAGES: libpython3-dev
+        RUNTIME_PACKAGES: libpython3.9 python3-numpy git
+        RUNTIME_PIP_PACKAGES: "--no-deps --no-use-pep517 git+https://github.com/lnls-sirius/dev-packages.git@master#egg=siriuspy&subdirectory=siriuspy git+https://github.com/lnls-fac/mathphys.git@master#egg=mathphys numpy==1.23"
```
I think you can split this into separate lines using `>`, as done in https://github.com/lnls-dig/uhal/blob/18f26f102f2ebed2ded3fc70fae77d782c813c41/.github/workflows/static-analysis.yml#L25

ericonr commented on iocBoot/iocsi_ps_conv_fastcorrs/st.cmd:
```diff
@@ -1,22 +1,15 @@
-#dbLoadRecords("db/cntrec.db","user=gustavoreis")
-dbLoadRecords("db/kicks.db","user=gustavoreis, P=$(P), R=$(R)")
+system("./generate_crate_substitutions.py")
+dbLoadTemplate("db/crates.substitutions", "TEST=test-")
```
We could use `on error break` or `halt` to catch any errors that might happen in the python script, like bad environment variable values.

ericonr replied:
Why not get `TEST` from the environment?

ericonr commented on generate_crate_substitutions.py:
```diff
@@ -0,0 +1,41 @@
+            start = int(start_str)
+            end = int(end_str)
+            for num in range(start, end + 1):
+                sections.append(num)
```
```suggestion
            sections.extend(range(start, end + 1))
```

ericonr commented on si_ps_conv_fastcorrsApp/Db/Makefile:
```diff
@@ -6,8 +6,9 @@ include $(TOP)/configure/CONFIG
 # Create and install (or just install) into <top>/db
 # databases, templates, substitutions like this
-#DB += cntrec.db
 DB += kicks.db
```
We don't need to install `kicks.db`, right?

gustavosr8 replied:
I don't fully understand it, but msi didn't like when it isn't installed

```
ERROR msi: Can't open file 'kicks.db'                                                                                                
input: '' at                                                                                                                         
Inflating database from ../dipole.substitutions                                                                                      
"/opt/epics/base/bin/linux-x86_64/msi"    -I. -I.. -I../O.Common -I../../../db -I/opt/epics/base/db -I/opt/epics/modules/pvxs/db -I/o
pt/epics/modules/sequencer/db -I/opt/epics/modules/calc/db -I/opt/epics/modules/asyn/db -I/opt/epics/modules/modbus/db -I/opt/epics/m
odules/busy/db -I/opt/epics/modules/autosave/db -I/opt/epics/modules/sscan/db -I/opt/epics/modules/recsync/db -I/opt/epics/modules/et
her_ip/db -I/opt/epics/modules/iocStats/db -I/opt/epics/modules/ipmiComm/db -I/opt/epics/modules/pyDevSup/db -I/opt/epics/modules/snm
p/db -I/opt/epics/modules/scaler/db -I/opt/epics/modules/mca/db -I/opt/epics/modules/areaDetector/ADAravis/db -I/opt/epics/modules/ar
eaDetector/ADGenICam/db -I/opt/epics/modules/areaDetector/ADSimDetector/db -I/opt/epics/modules/areaDetector/ADCore/db -I/opt/epics/m
odules/NDSSCPimega/db -I/opt/epics/modules/motor/db -I/opt/epics/modules/motor/modules/motorPIGCS2/iocs/pigcs2IOC/db -I/opt/epics/mod
ules/pmac/db -I/opt/epics/modules/opcua/db -o dipole.db -S../dipole.substitutions               
```

gustavosr8 replied:
Nevermind :sweat_smile: 
Fixed it!

ericonr commented on si_ps_conv_fastcorrsApp/Db/dipole.substitutions:
```diff
@@ -0,0 +1,12 @@
+file "kicks.db"
```
And this could use `kicks.template` directly.

Why did you call this file `dipole`, though? Shouldn't it be something like sector?

gustavosr8 replied:
I like this, gonna follow with `sector`!

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+    field(PREC, 3)
+    field(SCAN, "Passive")
+    field(FLNK, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB PP")
+}
```
Shouldn't `out` and `copy` records also have `$(TEST)` prefix?

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+record(ao, "out-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP") {
```
What do DIP and ORI stand for?

gustavosr8 replied:
`Dipole` and `Orientation`. Do you think that the complete names are a better approach?

ericonr replied:
Dipole is the wrong terminology. We have each ring sector (1 through 20), and they have "trechos" (M1, M2, ...). These can be related to a given dipole (e.g. C2 and C3-1 are around the BC dipole), but not necessarily (e.g. M1 and M2 are around the straight section).

Instead of Orientation, we usually use Plane.

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+    field(EGU, "urad")
+    field(PREC, 3)
+    field(DTYP, "Python Device")
+    field(OUT, "@strn2curr $(SEC) $(DIP) $(ORI) Current-SP")
```
If everything receives a `TEST` prefix, I suppose all `@` expressions would use `$(TEST)$(SEC)` for the first argument

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+    field(ONAM, "Ext PV OK")
+}
+
+record(sseq, "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP") {
+    field(PINI, "YES")
+    field(DOL1, "SI-$(SEC)$(DIP):PS-$(ORI):Current-SP CP")
+    field(LNK1, "copy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP PP")
+}
```
Either this record or `copy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP` are in the wrong spot, since these records depend on one another and should be easy to view simultaneously.

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+    field(FLNK, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB PP")
+}
+
+record(ai, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB") {
+    field(EGU, "urad")
+    field(PREC, 3)
+    field(DTYP, "Python Device")
+    field(INP, "@curr2strn $(SEC) $(DIP) $(ORI) Current-SP")
+    field(SCAN, "Passive")
+    field(FLNK, "$(TEST)fwd-SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB")
+}
+
+record(ao, "$(TEST)fwd-SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB") {
+    field(EGU, "urad")
+    field(PREC, 3)
+    field(DOL, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB")
+    field(OMSL, "closed_loop")
+    field(OUT, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):Kick-SP PP")
+}
```
It would be nice to have a comment like: "this is done to update `Kick-SP` when `Current-SP` is updated by someone else"

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+    field(EGU, "urad")
+    field(PREC, 3)
+    field(DTYP, "Python Device")
+    field(SCAN, "5 second")
```
I think a comment explaining why the scan is necessary would be nice

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,219 @@
+    field(OUT, "SI-$(SEC)$(DIP):PS-$(ORI):Kick-SP.DRVH")
+}
+
+record(sseq, "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):Current-RB") {
+    field(PINI, "YES")
+    field(DOL1, "SI-$(SEC)$(DIP):PS-$(ORI):Current-RB CP")
+    field(LNK1, "copy-SI-$(SEC)$(DIP):PS-$(ORI):Current-RB PP")
+}
+
+record(bi, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):CurrentRBCnct-Mon") {
+    field(PINI, "YES")
+    field(SCAN, "1 second")
+    field(INP , "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):Current-RB.DOL1V")
+    field(ZNAM, "Ext PV Not Connected")
+    field(ZSV , "MAJOR")
+    field(ONAM, "Ext PV OK")
+}
```
Now that we understand build-time generation of dbs better, why not have a template that works for all of `Current-RB/Mon`, `CurrentRef`, `FOFBAcc`? That way you need to specify the DB logic only once, and we can instantiate it multiple times

gustavosr8 committed:
Use template/substitutions to avoid repeated db logic.

In order for the substitution to work correctly, we need to remove
`Cnct` from the middle of the suffix. This is more relevant to the
kick_readback PVs. The setpoint ones are changed to maintain
consistency.

gustavosr8 committed:
Use common database pattern for kick and current setpoint sync.

Previously, we depended on the value of Current-SP eventually settling,
because the direct and inverse operations in use should converge onto a
specific Kick and Current pair. However, any update of Current-SP would
cause at least one write operation into Current-SP due to processing of
Kick-SP, including the update caused by a value change of Kick-SP. In
normal operating conditions, this is not a big issue, but, for this IOC,
when we don't have an electron beam and the dipoles are turned off (0
GeV beam energy), all current values are converted into 0 urad kicks,
which are then converted into 0 A currents. This makes it impossible to
operate the fast corrector power supplies, since this IOC is constantly
forcing the Current-SP values to 0 whenever a different value is written
into them.

Now, since we are using SDIS and PACT, we can specifically disable any
writes into Current-SP when updating Kick-SP based on it, but writes
into Kick-SP still update Current-SP accordingly.

[1] https://docs.epics-controls.org/en/latest/process-database/common-database-patterns.html#readback-from-a-different-hardware-address

gustavosr8 commented:
Missing new tests after the rework, with focus on the last two commits

ericonr commented on docker-compose.yml:
```diff
@@ -0,0 +1,23 @@
+        RUNTIME_PIP_PACKAGES: >
+          --no-deps --no-use-pep517
+          git+https://github.com/lnls-sirius/dev-packages.git@master#egg=siriuspy&subdirectory=siriuspy
+          git+https://github.com/lnls-fac/mathphys.git@master#egg=mathphys numpy==1.24
```
Could we choose specific releases of each, to avoid breaking the IOC on a simple rebuild?

And can you move `numpy` into its own line, to avoid confusion?

ericonr commented on generate_crate_substitutions.py:
```diff
@@ -0,0 +1,40 @@
+        f.write('}\n')
+
+        f.write('file "db/machine_params.db"\n')
+
```
I would simplify things by having the `sector.substitutions` file include `machine_params.db`

ericonr commented on si_ps_conv_fastcorrsApp/Db/kick_conv.template:
Nit: name this file something including "readback" in name?

e.g. `kick_readback_conv`

ericonr commented on si_ps_conv_fastcorrsApp/Db/kick_conv.template:
```diff
@@ -0,0 +1,29 @@
+    field(EGU, "urad")
+    field(PREC, 3)
+    field(DTYP, "Python Device")
+    field(INP, "@curr2strn $(TEST) $(SEC) $(DIP) $(ORI) $(CURRSUFFIX)")
```
What happens if the TEST prefix is empty?

Since you are the one doing the string splitting, I think the prefix is just going to be missing. You might need to think of optional arguments (different logic based on the length of the result of split)

ericonr commented on si_ps_conv_fastcorrsApp/Db/kick_curr_pairs.substitutions:
```diff
@@ -0,0 +1,10 @@
+}
+
+file "kick_limits.template"
+{ pattern { TYPE } { "DRVL" } { "DRVH" } }
```
I think this would benefit from the same formatting as the block above.

ericonr commented on si_ps_conv_fastcorrsApp/Db/kick_limits.template:
Nit: `kick_setpoint_limits.template`

ericonr commented on si_ps_conv_fastcorrsApp/Db/kick_limits.template:
```diff
@@ -0,0 +1,24 @@
+    field(EGU, "A")
+    field(PREC, 3)
+    field(INP, "SI-$(SEC)$(DIP):PS-$(ORI):Current-SP.$(TYPE) CP")
+    field(FLNK, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):Kick-$(TYPE) PP")
+}
+
+record(ai, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):Kick-$(TYPE)") {
+    field(EGU, "urad")
+    field(PREC, 3)
+    field(DTYP, "Python Device")
+    field(SCAN, "5 second")
+    field(INP, "@curr2strn $(TEST) $(SEC) $(DIP) $(ORI) Current-$(TYPE)")
+    field(FLNK, "$(TEST)fwd-SI-$(SEC)$(DIP):PS-$(ORI):Kick-$(TYPE)")
```
You're using `FLNK ... PP` for some links, but not all. Considering that it's not necessary, what do you think  about omitting `PP` for all cases?

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,56 @@
+include "kick_curr_pairs.db"
+
+# Records for sync Kick and Current setpoints
+record(bi, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):CnctCurrentSP-Mon") {
+    field(PINI, "YES")
+    field(SCAN, "1 second")
+    field(INP,  "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP.DOL1V" )
+    field(ZNAM, "Ext PV Not Connected")
+    field(ZSV , "MAJOR")
+    field(ONAM, "Ext PV OK")
+}
+
+record(sseq, "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP") {
+    field(PINI, "YES")
+    field(DOL1, "SI-$(SEC)$(DIP):PS-$(ORI):Current-SP CP")
+    field(LNK1, "$(TEST)copy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP PP")
+}
```
Use same order as `kick_conv`?

ericonr commented on si_ps_conv_fastcorrsApp/Db/kick_conv.template:
```diff
@@ -0,0 +1,29 @@
+record(sseq, "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):$(CURRSUFFIX)") {
+    field(PINI, "YES")
```
I don't think it makes a lot of sense to use `PINI=YES` in a record where the input record is `CP`. That means it will be processed immediately when the PV connects, and processing before that will use a zeroed out value.

ericonr commented on si_ps_conv_fastcorrsApp/src/strn2curr.py:
```diff
@@ -3,22 +3,28 @@
-        self.base_record.VAL = norm.conv_strength_2_current(rec.VAL, strengths_dipole=self.dipole_strehgth)
+        self.base_record.VAL = self.norm.conv_strength_2_current(
+            rec.VAL, strengths_dipole=self.dipole_strength.VAL)
+        self.base_record.scan(sync=True)
```
Do you think it could work to remove this scan, and use a `FLNK` in the record instead?

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -13,13 +16,11 @@ record(sseq, "$(TEST)proxy-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP") {
+    field(PREC, 3)
+    field(SCAN, "Passive")
+    field(FLNK, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB PP")
 }
```
Why did this commit change the order of these records?

gustavosr8 replied:
To be more logic to follow the processing chain

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -29,19 +30,20 @@ record(ao, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):Kick-SP") {
-    field(INP, "@curr2strn $(TEST) $(SEC) $(DIP) $(ORI) Current-SP")
     field(SCAN, "Passive")
+    field(INP, "@curr2strn $(TEST) $(SEC) $(DIP) $(ORI) Current-SP")
+    field(PINI, "YES")
```
How does PINI work here?

gustavosr8 replied:
Not sure, tbh. Just following the reference

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -0,0 +1,56 @@
+    field(EGU, "A")
+    field(PREC, 3)
+    field(OUT, "SI-$(SEC)$(DIP):PS-$(ORI):Current-SP CA")
+    field(SDIS, "fwd-SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB.PACT")
```
I think it would be nice to have a comment documenting the record processing chain and its usage of only local links, so that when this record is reached, KickSP-RB will, indeed, still be processing.

proxy-Current-SP -> copy-Current-SP -> KickSP-RB -> fwd-KickSP-RB -> Kick-SP -> out-Current-SP

(this is another reason for my suggestion of using a FLNK in Kick-SP, it makes the processing chain more explicit).

gustavosr8 committed:
Change SEC and DIP macros to match LNLS name convention.

DIP, that stands for "Dipole", was the wrong terminology. "Subsection"
(SUBSEC) is a more correct term. Same goes to Orientation (ORI), since
usually "Plane" is used.

gustavosr8 committed:
Add script for delay calculation between setpoint and readback.

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -1,16 +1,18 @@
 
-record(bi, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):CurrentSPCnct-Mon") {
+record(bi, "$(TEST)SI-$(SEC)$(DIP):PS-$(ORI):CnctCurrentSP-Mon") {
     field(PINI, "YES")
```
These changes sneaked into the wrong commit.

gustavosr8 replied:
I don't think they're in the wrong commit; they just aren't explained in the message. In order for the substitution to work correctly, we need to remove `Cnct` from the middle of the suffix. This is more relevant to the kick_readback PVs. The setpoint ones are changed to maintain consistency.

ericonr commented on si_ps_conv_fastcorrsApp/Db/kicks.template:
```diff
@@ -1,11 +1,9 @@
-    field(OUT, "SI-$(SEC)$(DIP):PS-$(ORI):Current-SP CA")
-}
+# The record processing chain here goes in this way:
+# proxy-Current-SP -> copy-Current-SP -> KickSP-RB -> fwd-KickSP-RB -> Kick-SP -> out-Current-SP
```
"The record processing chain here is the following:"

ericonr commented on jitter.py:
This is not exactly jitter; jitter is timing variation (so something like the standard deviation of the time between PV updates?)

I think we can be more accurate with the naming here.

I'd also appreciate it if the file had some additional comments; I could follow the logic flow, but I had to read the whole thing.

ericonr commented on jitter.py:
```diff
@@ -0,0 +1,178 @@
+import csv
+import threading
+from collections import defaultdict
+import sys
```
These last two imports don't seem to be used.

ericonr commented on jitter.py:
```diff
@@ -0,0 +1,178 @@
+                for pln in self.args.plane:
+                    cell_str = f"{sec:02d}"
+                    sp = f"{self.args.prefix}SI-{cell_str}{subsec}:PS-{pln}:Kick-SP"
+                    rb = f"{self.args.prefix}SI-{cell_str}{subsec}:PS-{pln}:Current-RB"
```
Hmm, this is interesting

You're measuring the time for the low level IOC to respond as well. Does that make sense? Is it so small that we can see a big enough difference between this IOC and the old one? We could instead look at two pairs, Kick-SP/Current-SP, and Current-RB/Kick-RB or Current-Mon/Kick-Mon, to try and isolate only the part pertaining to this IOC.

ericonr commented on jitter.py:
```diff
@@ -0,0 +1,178 @@
+
+            # Only record if value matches the test value
+            if abs(value - measurement['test_value']) < 0.0001:
+                measurement['sp_timestamp'] = time.time()
+                measurement['expecting_sp'] = False
+                measurement['expecting_rb'] = True
+
+    def on_readback_change(self, pvname, value, **kwargs):
+        with self.measurement_lock:
+            if pvname not in self.active_measurements:
+                return
+
+            measurement = self.active_measurements[pvname]
+            if not measurement['expecting_rb']:
+                return
+
+            measurement['rb_timestamp'] = time.time()
```
Should we use our client's time (subject to multiple kinds of uncertainties) or our PVs' timestamps? Or both?

ericonr commented on jitter.py:
```diff
@@ -0,0 +1,178 @@
+            epics.caput(sp, test_value, wait=False)
+
+            timeout_time = start_time + self.args.timeout
+            while time.time() < timeout_time:
+                with self.measurement_lock:
+                    if self.active_measurements[sp]['completed']:
+                        jitter = (self.active_measurements[sp]['rb_timestamp'] -
+                                  self.active_measurements[sp]['sp_timestamp'])
+                        jitters.append(jitter)
+                        break
+                time.sleep(0.001)
+            else:
+                jitters.append(None)
```
Couldn't this wait on a completed event in order to avoid busy looping?

ericonr commented on jitter.py:
```diff
@@ -0,0 +1,178 @@
+
+            epics.caput(sp, orig_sp, wait=True)
+
+            with self.measurement_lock:
+                if sp in self.active_measurements:
+                    del self.active_measurements[sp]
```
How can this member have already been deleted?

ericonr commented on jitter.py:
```diff
@@ -0,0 +1,178 @@
+            measurement['expecting_rb'] = False
+            measurement['completed'] = True
+
+    def measure_for_pv(self, sp, rb, sec, subsec, pln):
```
Am I correct in assuming this function is ready to be called by multiple different threads, but this isn't done in the current script?

gustavosr8 replied:
Yeap, the function is thread safe, but the code itself isn't using multiple threads. Could be done, if needed.

ericonr commented:
```
Change kick and current setpoint sync to use the tech-talk mechanism.

Listed in epics-controls 'common database patterns' as 'Readback from a
different hardware address' [1], we have a defined pattern that can be
followed in our case. It is similar to what was been doing already, with
the difference that we use the SDIS field of the
$(TEST)out-SI-$(SEC)$(DIP):PS-$(ORI):Current-SP record to listen to
$(TEST)fwd-SI-$(SEC)$(DIP):PS-$(ORI):KickSP-RB.PACT, in a way that, when
we have a current setpoint update, the "write" record processing is
disabled, preventing an infinite loop.

[1] https://docs.epics-controls.org/en/latest/process-database/common-database-patterns.html#readback-from-a-different-hardware-address
```

Use common database pattern for kick and current setpoint sync.

This design was based on a common database pattern from EPICS documentation [1].

Previously, we depended on the value of Current-SP eventually settling, because the direct and inverse operations in use should converge onto a specific Kick and Current pair. However, any update of Current-SP would cause at least one write operation into Current-SP due to processing of Kick-SP, including the update caused by a value change of Kick-SP. In normal operating conditions, this is not a big issue, but, for this IOC, when we don't have an electron beam and the dipoles are turned off (0 GeV beam energy), all current values are converted into 0 urad kicks, which are then converted into 0 A currents. This makes it impossible to operate the fast corrector power supplies, since this IOC is constantly forcing the Current-SP values to 0 whenever a different value is written into them.

Now, since we are using SDIS and PACT, we can specifically disable any writes into Current-SP when updating Kick-SP based on it, but writes into Kick-SP still update Current-SP accordingly.

---

**Note**: this was never mentioned on tech-talk; the link already lists the exact section

gustavosr8 committed:
Add script to test CPU overload on multiple setpoints.

ericonr commented on configure/CONFIG_SITE:
```diff
@@ -1,44 +1,5 @@
-#   or CROSS_OPT settings from base/configure/CONFIG_SITE
-#HOST_OPT = NO
-#CROSS_OPT = NO
 PY_VER ?= 3.9
```
can we remove this entry as well?

gustavosr8 replied:
Yeap, this was for one of the other tested modules.
